flows/contact_test.go

```diff
@@ -186,97 +189,51 @@ func TestContactSetPreferredChannel(t *testing.T) {
-	age := test.NewField("age", "Age", assets.FieldTypeNumber)
-
-	fieldSet := flows.NewFieldAssets([]assets.Field{gender.Asset(), age.Asset()})
-
-	males := test.NewGroup("Males", `gender="M"`)
-	old := test.NewGroup("Old", `age>30`)
-	english := test.NewGroup("English", `language=eng`)
-	spanish := test.NewGroup("Español", `language=spa`)
-	lastYear := test.NewGroup("Old", `created_on <= 2017-12-31`)
-	tel1800 := test.NewGroup("Tel with 1800", `tel ~ 1800`)
-	twitterCrazies := test.NewGroup("Twitter Crazies", `twitter ~ crazy`)
-	broken := test.NewGroup("Broken", `xyz="X"`) // no such field
-	groups := []*flows.Group{males, old, english, spanish, lastYear, tel1800, twitterCrazies, broken}
-
-	contact := flows.NewEmptyContact(session.Assets(), "Joe", "eng", nil)
-	contact.AddURN(urns.URN("tel:+12345678999"), nil)
-
-	memberships, errors := evaluateGroups(t, env, contact, groups, fieldSet)
-	assert.Equal(t, []*flows.Group{english}, memberships)
-	assert.Equal(t, []*flows.Group{broken}, errors)
-
-	contact.SetLanguage(envs.Language("spa"))
-	contact.AddURN(urns.URN("twitter:crazy_joe"), nil)
-	contact.AddURN(urns.URN("tel:+18005555777"), nil)
-
-	genderValue := contact.Fields().Parse(env, fieldSet, gender, "M")
-	contact.Fields().Set(gender, genderValue)
-
-	ageValue := contact.Fields().Parse(env, fieldSet, age, "37")
-	contact.Fields().Set(age, ageValue)
-
-	contact.SetCreatedOn(time.Date(2017, 12, 15, 10, 0, 0, 0, time.UTC))
+	tests := []struct {
+		Description   string          `json:"description"`
+		ContactBefore json.RawMessage `json:"contact_before"`
+		RedactURNs    bool            `json:"redact_urns"`
+		ContactAfter  json.RawMessage `json:"contact_after"`
+	}{}
 
-	memberships, errors = evaluateGroups(t, env, contact, groups, fieldSet)
-	assert.Equal(t, []*flows.Group{males, old, spanish, lastYear, tel1800, twitterCrazies}, memberships)
-	assert.Equal(t, []*flows.Group{broken}, errors)
-}
-
-func TestReevaluateDynamicGroupsWithURNRedaction(t *testing.T) {
-	session, _, err := test.CreateTestSession("http://localhost", envs.RedactionPolicyURNs)
+	testFile, err := ioutil.ReadFile("testdata/dynamic_groups.json")
+	require.NoError(t, err)
+	err = jsonx.Unmarshal(testFile, &tests)
 	require.NoError(t, err)
 
-	env := session.Runs()[0].Environment()
-
-	gender := test.NewField("gender", "Gender", assets.FieldTypeText)
-	age := test.NewField("age", "Age", assets.FieldTypeNumber)
-
-	fieldSet := flows.NewFieldAssets([]assets.Field{gender.Asset(), age.Asset()})
-
-	males := test.NewGroup("Males", `gender="M"`)
-	tel1800 := test.NewGroup("Tel with 1800", `tel ~ 1800`)
-	twitterCrazies := test.NewGroup("Twitter Crazies", `twitter ~ crazy`)
-	groups := []*flows.Group{males, tel1800, twitterCrazies}
-
-	contact := flows.NewEmptyContact(session.Assets(), "Joe", "eng", nil)
-	contact.AddURN(urns.URN("tel:+12345678999"), nil)
+	for _, tc := range tests {
+		sa, err := engine.NewSessionAssets(source, nil)
```

Comment:
Rewriting these tests led me to realize something odd - session assets aren't tied to particular environment - I tried to use the same session assets for each test case.. but.. the parsed contact query of a group is cached on the group. and it is environment specific

That led me to try reworking the whole assets creation thing so that session assets are created with an environment, query parsing happens at group creation (no caching needed)... but then it wasn't obvious how that parse error gets back to the user... so reverted for now.

Reply:
Is there ever a time in the real world where an environment is different for the same session assets? Not sure I understand. It is very very important to mailroom that Session assets are both thread safe and able to be reused on the same org so we should definitely discuss if that isn't the case.

Reply:
No.. this isn't something we do. But it does kinda annoy me that it's possible, but subtly doesn't work correctly because of the cached contact query. Would be nice to get rid of that little bit of caching.